### PR TITLE
Add Geometry buffer_with_style and buffer_with_params Geos API support.

### PR DESCRIPTION
**Pull request by enerqi**

### CODE REVIEW

File: src/buffer_params.rs [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+use crate::{
+    AsRaw, AsRawMut, ContextHandle, ContextHandling, ContextInteractions, Error, GResult, JoinStyle,
+};
+
+use geos_sys::*;
+use std::sync::Arc;
+
+pub const DEFAULT_MITRE_LIMIT: f64 = 5.0;
+pub const DEFAULT_QUADRANT_SEGMENTS: i32 = 8;
+pub const DEFAULT_SINGLE_SIDED: bool = false;
```
Comment (GuillaumeGomez) [inline src/buffer_params.rs:12]: That seems like they should all be enums.

Comment (enerqi) [reply src/buffer_params.rs:12]: You would have to explain - I can't see how an enum would fit here. 
The constants are already namespaced by the `buffer_params` module. If those constants became an enum we would then have to create a separate instance of it as the individual cases can't have defaults (a single value) without being a C style enum. Even then a C-style enum is not well typed unless all cases are integers. In usage, such as `buffer_with_style` the cases are not exclusive/disjoint. The C API has separate set param functions (mitre, cap style...) and not a single setParam that can take a union of inputs.

Comment (GuillaumeGomez) [reply src/buffer_params.rs:12]: Because I assume there are other constants like `*_MITRE_*` and for others too.

Comment (enerqi) [reply src/buffer_params.rs:12]: You want the enum to behave like a "constants" namespace instead of just using the buffer params module as the namespace? Seems like the actual constant definitions would then have to be a hashmap mapping from the constant type mitre | quad segments | etc to the constant value.

File: src/wkb_writer.rs [hunk 2]
```diff
@@ -7,7 +7,7 @@ use geos_sys::*;
 use std::convert::TryFrom;
 use std::sync::Arc;
 
-/// The `WKBWriter` type is used to generate `HEX` or `WKB` formatted output from [`Geometry`].
+/// The `WKBWriter` type is used to generate `HEX` or `WKB` formatted output from [`Geometry`](crate::Geometry).
```
Comment (GuillaumeGomez) [inline src/wkb_writer.rs:10]: Why this change?

File: src/prepared_geometry.rs [hunk 3]
```diff
@@ -30,7 +30,7 @@ pub struct PreparedGeometry<'a> {
 }
 
 impl<'a> PreparedGeometry<'a> {
-    /// Creates a new `PreparedGeometry` from a [`Geometry`].
+    /// Creates a new `PreparedGeometry` from a [`Geometry`](crate::Geometry).
```
Comment (GuillaumeGomez) [inline src/prepared_geometry.rs:33]: Why this change?

File: src/wkb_writer.rs [hunk 4]
```diff
@@ -202,8 +202,8 @@ impl<'a> WKBWriter<'a> {
         }
     }
 
-    /// Returns the number of dimensions to be used when calling [`WKBWriter::write`]. By default,
-    /// it is 2.
+    /// Returns the number of dimensions to be used when calling [`WKBWriter::write_wkb`].
```
Comment (GuillaumeGomez) [inline src/wkb_writer.rs:205]: Why this change?

File: src/wkt_writer.rs [hunk 5]
```diff
@@ -9,7 +9,7 @@ use geos_sys::*;
 use std::convert::TryFrom;
 use std::sync::Arc;
 
-/// The `WKTWriter` type is used to generate `WKT` formatted output from [`Geometry`].
+/// The `WKTWriter` type is used to generate `WKT` formatted output from [`Geometry`](crate::Geometry).
```
Comment (GuillaumeGomez) [inline src/wkt_writer.rs:12]: Why this change?

File: src/buffer_params.rs (hunk 1 above)
Comment (GuillaumeGomez) [reply src/buffer_params.rs:12]: Ok, so after taking a look around, I'm not sure if we should keep these constants... If we want to handle default values, might be better to implement `Default` on `BufferParams` instead, no?

File: src/wkb_writer.rs (hunk 2 above)
Comment (enerqi) [reply src/wkb_writer.rs:10]: All these doc string changes are because the links are broken (Geometry is not actually imported in that module), e.g. without the addition when running cargo doc:

> warning: unresolved link to `Geometry`
>   --> src\wkb_writer.rs:10:85
>    |
> 10 | /// The `WKBWriter` type is used to generate `HEX` or `WKB` formatted output from [`Geometry`]
>    |                                                                                     ^^^^^^^^ no item named `Geometry` in scope
>    |
>    = note: `#[warn(rustdoc::broken_intra_doc_links)]` on by default
>    = help: to escape `[` and `]` characters, add '\' before them like `\[` or `\]`

Comment (GuillaumeGomez) [reply src/wkb_writer.rs:10]: I see, that's weird it wasn't seen before. Thanks for the fix!

File: src/buffer_params.rs (hunk 1 above)
Comment (enerqi) [reply src/buffer_params.rs:12]: I can change the `BufferParamsBuilder` to use options instead of default values and only call the C Api to set the values if the option is `Some`...and remove those constants. 
`BufferParams` doesn't really need default as it has `new`. Not setting any options will use the defaults defined in the  C++ API.

Comment (GuillaumeGomez) [reply src/buffer_params.rs:12]: Fine by me!

**Review by GuillaumeGomez — APPROVED** [review]
